scrapy_poet/injection.py

```diff
@@ -246,7 +246,37 @@ def build_callback_dependencies(self, request: Request, response: Response):
         """
         plan = self.build_plan(request)
         provider_instances = yield from self.build_instances(request, response, plan)
-        return plan.final_kwargs(provider_instances)
+        final_kwargs = plan.final_kwargs(provider_instances)
+        final_kwargs = yield self.convert_po_into_item(request, final_kwargs)
+        return final_kwargs
+
+    @inlineCallbacks
+    def convert_po_into_item(self, request: Request, final_kwargs: dict):
+        """Given a mapping of callback dependencies, call the ``to_item()``
+        method of a dependency to return the requested item.
+        """
+        callback_deps = andi.inspect(get_callback(request, self.spider))
```

Comment:
hey @BurnzZ! That's an interesting implementation, I haven't thought about it. 
Would it work e.g. to provide Product as a dependency to a Page Object?
I was thining we're going to have a provider for item classes.

Reply:
Hi @kmike , I've attempted to move this logic into a new provider.

For example, let's have the following setup:

```python
@attrs.define
class InjectedInfo(ItemFromPageObject):
    info: str

@handle_urls(URL)
class InjectedInfoPage(ItemPage[ItemFromPageObject]):
    @field
    def info(self) -> str:
        return "injected info"

@attrs.define
class Product:
    name: str
    injected_info: InjectedInfo

@handle_urls(URL)
@attrs.define
class ProductPage(ItemPage[Product]):
    injected_info_item: InjectedInfo

    @field
    def name(self) -> str:
        return "product name"

    @field
    def injected_info(self) -> InjectedInfo:
        return self.injected_info_item
```

Using the above example, we get `to_provide=InjectedInfo` inside the provider.

The next step involves trying to find the PageObject that is able to return the item (to be provided) for the given request URL. This logic is quite straightforward and we could get `InjectedInfoPage` without any problems.

However, the derived `InjectedInfoPage` Page Object is still a class and yet to be instantiated. We can't directly instantiate it within the provider (*and call* `to_item()`) since the Page Object class could have its own dependencies (e.g. `HttpResponse`, `HttpClient`, etc) that needs to be resolved first.

In particular, the `InjectedInfoPage` *could be fully built* on this portion of the **injector.py** module:

```python
def build_instances(self, request: Request, response: Response, plan: andi.Plan):
    # First we build the external dependencies using the providers
    instances = yield from self.build_instances_from_providers(
        request, response, plan
    )

    # All the remaining dependencies are internal so they can be built just
    # following the andi plan.
    for cls, kwargs_spec in plan.dependencies:
        if cls not in instances.keys():
            instances[cls] = cls(**kwargs_spec.kwargs(instances))
    return instances
```

However, `InjectedInfoPage` is not even in the `andi.Plan` yet in the code snippet above. In order to make providers work, I'm considering refactoring **injector.py** such that we're able to make it properly work for our new use case.

Before that I wanted to ask if you see another way of solving this? as I might be missing some key concept in **andi** that could solve this easily.

-----

Another thing, I think we need to define in **web-poet** a marker for items signifying that a Page Object returns them. For example, observe that the `InjectedInfo` inherits from `ItemFromPageObject` from the above examples.

This enables the provider to easily identify if it should handle the dependency or not. For example:

```python
class ItemProvider(PageObjectInputProvider):
    provided_classes = lambda type_: isclass(type_) and issubclass(type_, ItemFromPageObject)
    name = "item"

    # and so on ...
```

However, I'm not to keen on this approach since it creates a poor user experience. There's a need to keep in mind to properly subclass from `ItemFromPageObject`, otherwise, it wouldn't be injected to the PageObject.

An alternative way is by checking out all the rules from the registry to see if the `type_` is indeed declared as `to_return` in any of the rules. However, this could be slow as there could be thousands of rules and we need to check each one of them. The `OverridesRegistry` needs to be refactored to support quick lookups like this.

Reply:
Hey!

1. I think you're right, and we'd need to do at least 2 iterations of injection to provide an item. I don't see an easy way around it.
2. I think we should use registry. Can it be called once though, why is the speed a issue? We get all rules with `to_return` which is not None, and that's our `provided_classes`.

Reply:
Updated the implementation to create an `ItemProvider`.